### PR TITLE
Iris to GeoVista conversion

### PR DESCRIPTION
**Pull request by HGWright**
### Description

From the issue #5517. This PR adds the ability to convert an Iris cube into a polydata mesh and then extract a region of that to be able to be plotted using geovista. This is a one way operation as it does not include the ability to go the other way from geovista to iris. There was plenty of discussion of the scope of this PR in the original issue.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        pytest.raises(
+            NotImplementedError,
+            match=r"Only 1D and 2D coordinates are supported",
+        )
+
+    def test_no_mesh_or_2d(self):
+        pytest.raises(
+            NotImplementedError,
+            match=r"Cube must have a mesh or have 2 dimensions",
+        )
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py:165]: None of these are doing anything, so they look like they're passing.

`pytest.raises` is used as a context manager:

```python
with pytest.raises(ValueError, match="foo"):
    my_operation()
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Good stuff @HGWright, here are some things to action when you have time

File: lib/iris/experimental/geobridge.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        lons, lats = cube.mesh.node_coords
+        face_node = cube.mesh.face_node_connectivity
+        indices = face_node.indices_by_location()
+
+        polydata = Transform.from_unstructured(
+            lons.points,
+            lats.points,
+            indices,
+            data=cube.data,
+            name=f"{cube.name()} / {cube.units}",
```
Comment (trexfeathers) [inline lib/iris/experimental/geobridge.py:27]:
```suggestion
            name=f"{cube.name()} / ({cube.units})",
```

I just checked what it looks like when printing a `Cube`.

File: lib/iris/experimental/geobridge.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            start_index=face_node.start_index,
+        )
+    elif cube.ndim == 2:
+        x_coord = _get_coord(cube, "X")
+        y_coord = _get_coord(cube, "Y")
+        kwargs = dict(
+            xs=x_coord.contiguous_bounds(),
+            ys=y_coord.contiguous_bounds(),
+            data=cube.data,
+            name=f"{cube.name()} / {cube.units}",
```
Comment (trexfeathers) [inline lib/iris/experimental/geobridge.py:37]:
```suggestion
            name=f"{cube.name()} / ({cube.units})",
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]

File: lib/iris/experimental/geobridge.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+def _get_coord(cube, axis):
+    try:
+        coord = cube.coord(axis=axis, dim_coords=True)
+    except CoordinateNotFoundError:
+        coord = cube.coord(axis=axis)
+    return coord
+
+
+def cube_faces_to_polydata(cube):
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/experimental/geobridge.py:16]: Just discovered that I NEED `zlevel` to get [this plotting example](https://scitools-iris.readthedocs.io/en/stable/further_topics/ugrid/operations.html#plotting) to work; otherwise the face cuts through the sphere of the globe:

image

File: lib/iris/experimental/geovista.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            recreate_mesh = True
+        elif cube.location == "node":
+            polydata_length = polydata.GetNumberOfPoints()
+            indices_key = VTK_POINT_IDS
+        else:
+            raise NotImplementedError("Must be on face or node.")
+
+        if cube.shape[mesh_dim] != polydata_length:
+            raise ValueError(
+                "The mesh on the cube and the polydata must have the" " same shape."
```
Comment (trexfeathers) [inline lib/iris/experimental/geovista.py:103]: Something weird has happened with the quotes here.

Anyway we discussed making this error maximally descriptive - telling the user the respective dimension lengths that Iris has found.

File: lib/iris/experimental/geovista.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+        if cube.location == "face":
+            polydata_length = polydata.GetNumberOfCells()
+            indices_key = VTK_CELL_IDS
+            recreate_mesh = True
+        elif cube.location == "node":
+            polydata_length = polydata.GetNumberOfPoints()
+            indices_key = VTK_POINT_IDS
+        else:
+            raise NotImplementedError("Must be on face or node.")
```
Comment (trexfeathers) [inline lib/iris/experimental/geovista.py:99]: A more descriptive error please. Should include what the actual value of `cube.location` is.

File: lib/iris/tests/integration/experimental/test_cube_to_poly.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+def test_integration_2d():
+    pass
+
+
+def test_integration_1d():
+    file_path = get_data_path(
+        [
+            "NetCDF",
+            "unstructured_grid",
+            "lfric_surface_mean.nc",
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/test_cube_to_poly.py:18]: Pretty sure this file also describes a mesh? You need a file with standard 1d coordinates (rectilinear coordinates).

File: lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        return {
+            "xs": cube_2d.coord(axis="X").contiguous_bounds(),
+            "ys": cube_2d.coord(axis="Y").contiguous_bounds(),
+            "data": cube_2d.data,
+            "name": cube_2d.name() + " / " + str(cube_2d.units),
+        }
+
+    @pytest.fixture()
+    def operation(self):
+        return "from_2d"
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py:112]: Not using this anymore

File: lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+            "start_index": 0,
+        }
+
+    @pytest.fixture()
+    def operation(self):
+        return "from_unstructured"
+
+    @pytest.fixture()
+    def cube(self, cube_mesh):
+        return cube_mesh
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py:160]: Not using this anymore

File: lib/iris/experimental/geovista.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        Incoming cube containing the arrays or the mesh to be converted into the
+        polydata object.
+
+    **kwargs : dict
+        Additional keyword arguments to be passed to the Transform method.
+
+    """
+    if cube.mesh:
+        if cube.ndim != 1:
+            raise NotImplementedError("Cubes with a mesh must be one dimensional")
```
Comment (trexfeathers) [inline lib/iris/experimental/geovista.py:39]: This needs test coverage

File: lib/iris/experimental/geovista.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+        else:
+            raise NotImplementedError("Only 1D and 2D coordinates are supported")
+    else:
+        raise NotImplementedError("Cube must have a mesh or have 2 dimensions")
+
+    return polydata
+
+
+def region_extraction(cube, polydata, region, **kwargs):
```
Comment (trexfeathers) [inline lib/iris/experimental/geovista.py:81]: Need to be consistent with the rest of Iris and make this an 'active' name e.g. `extract_region()`. But even that isn't fully descriptive; perhaps `extract_unstructured_cube_region()` or `extract_unstructured_region()`?

File: lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the BSD license.
+# See LICENSE in the root of the repository for full licensing details.
+"""Unit tests for the `iris.experimental.geobridge.cube_faces_to_polydata` function."""
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geobridge/test_cube_faces_to_polydata.py:5]:
```suggestion
"""Unit tests for the `iris.experimental.geovista.cube_faces_to_polydata` function."""
```

File: lib/iris/tests/unit/experimental/geobridge/test_region_extraction.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the BSD license.
+# See LICENSE in the root of the repository for full licensing details.
+"""Unit tests for the `iris.experimental.geobridge.region_extraction` function."""
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geobridge/test_region_extraction.py:5]:
```suggestion
"""Unit tests for the `iris.experimental.geovista.region_extraction` function."""
```

File: lib/iris/tests/unit/experimental/geobridge/__init__.py
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geobridge/__init__.py:1]: Directory needs to be `geovista` not `geobridge`.

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Some more comments. Nothing big!

File: lib/iris/experimental/geovista.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        if cube.location == "face":
+            polydata_length = polydata.GetNumberOfCells()
+            indices_key = VTK_CELL_IDS
+            recreate_mesh = True
+        elif cube.location == "node":
+            polydata_length = polydata.GetNumberOfPoints()
+            indices_key = VTK_POINT_IDS
+        else:
+            raise NotImplementedError(
+                f"Must be on face or node. Found: {cube.location}."
```
Comment (trexfeathers) [inline lib/iris/experimental/geovista.py:83]: What is it that `Must be on face or node`?

We need to empathise with the user here; they have run `extract_unstructured_region(...)` and received this message, they need a certain amount of information to be able to correct their mistake.

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+def test_integration_2d():
+    file_path = get_data_path(
+        [
+            "NetCDF",
+            "ORCA2",
+            "votemper.nc",
+        ]
+    )
+    with PARSE_UGRID_ON_LOAD.context():
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:23]: Don't need the context manager since this `Cube` is not UGRID.

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+def test_integration_1d():
+    file_path = get_data_path(
+        [
+            "NetCDF",
+            "global",
+            "xyt",
+            "SMALL_hires_wind_u_for_ipcc4.nc",
+        ]
+    )
+    with PARSE_UGRID_ON_LOAD.context():
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:41]: Don't need the context manager since this `Cube` is not UGRID.

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        global_cube = load_cube(file_path)
+
+    polydata = cube_to_polydata(global_cube[0, :])
+
+    assert polydata.GetNumberOfCells() == 51200
+    assert polydata.GetNumberOfPoints() == 51681
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:47]: We should also check the data, like `test_integration_mesh()`.

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+            "votemper.nc",
+        ]
+    )
+    with PARSE_UGRID_ON_LOAD.context():
+        global_cube = load_cube(file_path, "votemper")
+
+    polydata = cube_to_polydata(global_cube[0, 1, :])
+
+    assert polydata.GetNumberOfCells() == 26640
+    assert polydata.GetNumberOfPoints() == 26969
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:29]: We should also check the data, like `test_integration_mesh()`.

File: lib/iris/tests/integration/experimental/geovista/test_extract_unstructured_region.py [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        global_cube = load_cube(file_path, "conv_rain")
+    polydata = cube_to_polydata(global_cube[0, :])
+    region = BBox(lons=[0, 70, 70, 0], lats=[-25, -25, 45, 45])
+
+    extracted_cube = extract_unstructured_region(
+        global_cube, polydata, region, preference="center"
+    )
+
+    assert extracted_cube.ndim == 2
+    assert extracted_cube.shape == (72, 101)
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_extract_unstructured_region.py:34]: I'd like to see some assertions against `global_cube` too, to confirm what has and hasn't happened during extraction.

File: lib/iris/tests/unit/experimental/geovista/test_extract_unstructured_region.py [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+            polydata_length = self.mocked_polydata.GetNumberOfCells()
+        elif self.cube_mesh.location == "node":
+            polydata_length = self.mocked_polydata.GetNumberOfPoints()
+        with pytest.raises(
+            ValueError,
+            match=f"The mesh on the cube and the polydata"
+            f"must have the same shape."
+            f" Found Mesh: {self.cube_mesh.shape[self.cube_mesh.mesh_dim()]},"
+            f" Polydata: {polydata_length}.",
+        ):
```
Comment (trexfeathers) [inline lib/iris/tests/unit/experimental/geovista/test_extract_unstructured_region.py:149]: Nice dedication here! For future reference: the `match=` condition can just be a section of the error. We use that a lot when the error has dynamic content and we want to avoid the effort that you have put in here.

No change required though 🙂

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Nice work, here are my comments

File: docs/src/further_topics/ugrid/operations.rst [hunk 21]
```diff
@@ -488,7 +488,7 @@ earlier:
 
 
         # We'll re-use this to plot some real global data later.
-        >>> def cube_faces_to_polydata(cube):
+        >>> def cube_to_polydata(cube):
```
Comment (pp-mo) [inline docs/src/further_topics/ugrid/operations.rst:491]: Excuse a drive-by review, but shouldn't you now be replacing this definition with an import of the version from `iris.geovista` ?

Comment (trexfeathers) [reply docs/src/further_topics/ugrid/operations.rst]: Patience please, we're working on it

**Comment by pp-mo** [comment]
BTW there also seems to be nothing in the API docs build for `iris.geovista`
i.e. [here](https://scitools-iris--5740.org.readthedocs.build/en/5740/generated/api/iris.experimental.geovista.html)
I've no idea why this is -- I'm assuming it _should_ be showing API for `cube_to_polydata` and `extract_unstructured_region`.

**Comment by trexfeathers** [comment]
> BTW there also seems to be nothing in the API docs build for `iris.geovista` i.e. [here](https://scitools-iris--5740.org.readthedocs.build/en/5740/generated/api/iris.experimental.geovista.html) I've no idea why this is -- I'm assuming it _should_ be showing API for `cube_to_polydata` and `extract_unstructured_region`.

There's loads!

**Comment by pp-mo** [comment]
> > BTW there also seems to be nothing in the API docs build for `iris.geovista` i.e. [here](https://scitools-iris--5740.org.readthedocs.build/en/5740/generated/api/iris.experimental.geovista.html) I've no idea why this is -- I'm assuming it _should_ be showing API for `cube_to_polydata` and `extract_unstructured_region`.
> 
> There's loads!

Sorry, that was a browser caching thing !

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+import numpy as np
+
+from iris import load_cube
+from iris.experimental.geovista import cube_to_polydata
+from iris.experimental.ugrid import PARSE_UGRID_ON_LOAD
+from iris.tests import get_data_path
+
+
+def test_integration_2d():
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:15]: I've done some rough plotting to demonstrate this is doing the right thing:

### GeoVista

Screenshot 2024-03-27 145003




### Matplotlib (for comparison)

Screenshot 2024-03-27 150126

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    with PARSE_UGRID_ON_LOAD.context():
+        global_cube = load_cube(file_path, "votemper")
+
+    polydata = cube_to_polydata(global_cube[0, 1, :])
+
+    assert polydata.GetNumberOfCells() == 26640
+    assert polydata.GetNumberOfPoints() == 26969
+
+
+def test_integration_1d():
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:32]: I've done some rough plotting to demonstrate this is doing the right thing:

### GeoVista

Screenshot 2024-03-27 151038

### Matplotlib (for comparison)

Screenshot 2024-03-27 150823

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+    with PARSE_UGRID_ON_LOAD.context():
+        global_cube = load_cube(file_path)
+
+    polydata = cube_to_polydata(global_cube[0, :])
+
+    assert polydata.GetNumberOfCells() == 51200
+    assert polydata.GetNumberOfPoints() == 51681
+
+
+def test_integration_mesh():
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:50]: I've done some rough plotting to demonstrate this is doing the right thing:

Screenshot 2024-03-27 151500

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py (hunk 17 above)
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:47]: Please could we include a comment that indicates this is a 'known good output'?

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+        ]
+    )
+
+    with PARSE_UGRID_ON_LOAD.context():
+        global_cube = load_cube(file_path, "conv_rain")
+
+    polydata = cube_to_polydata(global_cube[0, :])
+
+    assert polydata.GetNumberOfCells() == 864
+    assert polydata.GetNumberOfPoints() == 866
```
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:65]: Please could we include a comment that indicates this is a 'known good output'?

File: lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py (hunk 18 above)
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/geovista/test_cube_to_poly.py:29]: Please could we include a comment that indicates this is a 'known good output'?

**Review by trexfeathers — CHANGES_REQUESTED** [review]

File: docs/src/further_topics/ugrid/operations.rst (hunk 21 above)
Comment (trexfeathers) [reply docs/src/further_topics/ugrid/operations.rst]: 16ef5216c7c1de0a5f212978d968749a878cb8cc

**Review by trexfeathers — APPROVED** [review]

**Comment by trexfeathers** [comment]
Thanks for sticking with it!
